tempodb/sharding_block_selector.go

```diff
@@ -0,0 +1,264 @@
+	return c.hash
+}
+
+func (c *shardedCompaction) CutBlock(currBlock *backend.BlockMeta, nextTraceID common.ID) bool {
+	if c.maxCompactionObjects > 0 && currBlock.TotalObjects >= c.maxCompactionObjects {
+		return true
+	}
+
+	if c.split {
+		// If the shard of the next trace is higher, then it's time to cut the block.
+		if shard := c.shardOf(nextTraceID); shard > c.shard {
+			c.shard = shard
+			return true
+		}
+	}
+
+	return false
+}
+
+var (
+	_ (common.CompactionRound)  = (*shardedCompaction)(nil)
+	_ (CompactionBlockSelector) = (*shardingBlockSelector)(nil)
+)
+
+func newShardingBlockSelector(shards int, blocklist []*backend.BlockMeta, maxCompactionRange time.Duration, maxCompactionObjects int, maxBlockBytes uint64, minInputBlocks, maxInputBlocks int) CompactionBlockSelector {
+	s := &shardingBlockSelector{
+		MinInputBlocks:       minInputBlocks,
+		MaxInputBlocks:       maxInputBlocks,
+		MaxCompactionRange:   maxCompactionRange,
+		MaxCompactionObjects: maxCompactionObjects,
+		MaxBlockBytes:        maxBlockBytes,
+		boundaries:           traceidboundary.All(uint32(shards)),
+	}
+
+	var (
+		now        = time.Now()
+		currWindow = s.windowForTime(now)
+	)
+
+	for _, b := range blocklist {
+
+		var (
+			w     = s.windowForBlock(b)
+			shard = s.shardOf(b.MinID)
+
+			// These are all of the numeric values that we can group and order by,
+			shardMin = fmt.Sprintf("%03d", shard)
+			shardMax = fmt.Sprintf("%03d", s.shardOf(b.MaxID))
+			columns  = fmt.Sprintf("%016X", b.DedicatedColumnsHash())
+			level    = fmt.Sprintf("%03d", b.CompactionLevel)
+			window   = fmt.Sprintf("%v", w)
+			age      = fmt.Sprintf("%016X", currWindow-w)
+			min      = fmt.Sprintf("%X", b.MinID)
+		)
+
+		entry := shardingBlockEntry{
+			meta:      b,
+			minBlocks: s.MinInputBlocks,
+			// Within group order by lowest compaction level first,
+			// then lowest trace ID to try to co-locate traces even further.
+			// This is the same for all cases.
+			order: join(level, min),
+		}
+
+		switch {
+		case b.CompactionLevel > 0 && shardMin == shardMax:
+			// This is a previously split block that contains only a single shard.
+			// Combine/dedupe with other blocks in the same window and shard.
+			// This is prioritized over splitting new blocks to keep the block count down.
+			entry.group = join(
+				"A",            // Highest priority
+				age,            // prioritize newer windows (more recent data)
+				shardMin,       // same shard
+				b.Version,      //
+				b.DataEncoding, // same block format and column config
+				columns,        //
+			)
+			entry.hash = join(
+				b.TenantID, // Work sharding by tenant, window
+				window,     //
+				shardMin,   // and shard
+			)
+
+		case b.CompactionLevel == 0:
+			// Unsharded new block that needs to be split. This step generates work
+			// for the above step. MinBlocks=1 allows stray single blocks to still be split.
+			entry.split = true
+			entry.minBlocks = 1
```

Comment:
wdyt about trying 2 (or more) input blocks at compaction 0 split into the shards?

Reply:
That was the original behavior but if your ingesters flush an odd number it will leave behind unsplit blocks.   It happens more than expected after testing internally.